Decoder metadata interface

197g opened the pull request:
See https://github.com/image-rs/image/issues/2245, the intended `ImageDecoder` changes.

This changes the `ImageDecoder` trait to fix some underlying issues. The main change is a clarification to the responsibilities; the trait is an interface from an implementor _towards the `image` library_. That is, the protocol established from its interface should allow _us_ to drive the decoder into our buffers and our metadata. It is _not_ optimized to be used by an external caller which should prefer the use of `ImageReader` and other inherent methods instead.

This is a work-in-progress, below motivates the changes and discusses open points.

- `ImageDecoder::peek_layout` encourages decoders to read headers after the constructor. This fixes the inherent problem we had with communicating limits. The sequences for internal use is roughly:
  ```rust
  let mut decoder = …;
  decoder.set_limits(…); // Other global configuration we have?

  { // Potentially multiple times:
    let layout_info = decoder.peek_layout()?;
    let mut buffer = allocate_for(&layout_info);
    decoder.assign_metadata()?;
    decoder.read_image(&mut buffer)?;
  }

  // … for sequences, start again from `peek_layout()`
  ```
- `ImageDecoder::read_image(&mut self)` no longer consumes `self`. We no longer need the additional `boxed` method and its trait work around, the trait is now dyn-compatible.

## Discussion

- [x] Maybe ~~`init`~~ `peek_layout` should return the full layout information in a single struct. We have a similar open issue for `png` in its own crate, and the related work for `tiff` is in the pipeline where its [`BufferLayoutPreference`](https://docs.rs/tiff/latest/tiff/decoder/struct.BufferLayoutPreference.html) already exists to be extended with said information.
  - [x] ~~Review limits and remove its size bounds insofar as they can be checked against the communicated bounds in the metadata step by the `image` side.~~ see: #2709, #2708
  - Idea: If a decoder supports builtin transforms (e.g. YCbCr -> Rgb conversion, grayscale, thumbnailing) that are more efficient than post-processing then there could be a negotiation phase here where information is polled twice / multiple times by different methods. The design should leave this negative space to be added in `1.1`, but it's not highly critical.
- [x] Fix the sequence decoder to use the new API
- [x] Tests for reading an image with `read_image` then switching to a sequence reader. But that is supposed to become mainly an adapter that implements the iterator protocol.
- [x] Remove remnants of the dyn-compatibility issue.
- [x] Adapt to the possibility of fetching metadata after the image. This includes changing `ImageReader` with a new interface to return some of it. That may be better suited for a separate PR though.
  - [x] Extract the CICP part of the metadata as `CicpRgb` and apply it to a decoded `DynamicImage`.
  - [x] Ensure that this is supported by all the bindings.
- [x] Deal with limits: https://github.com/image-rs/image/pull/2672#discussion_r2595580766

## Cleanup

- [ ] Better errors for `peek_layout` more consistently after `read_image`
  - [ ] avif
  - [ ] tga
  - [ ] pnm
  - [ ] tiff
  - [ ] dxt
  - [ ] qoi
  - [ ] dds
  - [ ] gif
- [ ] Make sure that `read_image` is 'destructive' in all decoders, i.e. re-reading an image and reading an image before `init` should never access an incorrect part of the underlying stream but instead return an error. Affects pnm and qoi for instance where the read will interpret bytes based on the dimensions and color, which would be invalid before reading the header and only valid for _one_ read.

197g committed:
Restructure ImageDecoder

Change ImageDecoder::read_image to `fn(&mut self)`. There are some
codecs where metadata and other information is only available after
decoding.

Add ImageDecoder::init for codecs where reading the header needs to take
into account limits and other configuration we are to introduce.

This sketch compiles as: `--no-default-features --features=png`

197g committed:
Migrate all codecs to init with layout

Renames it `peek_layout` to clarify the relation to `read_image`, that
it is to be called potentially multiple times for each such call.
Document the trait more with this relationship, moving some methods
around to clarify.

197g committed:
Ensure ICO forwards all methods

This was one of the silently broken features when adding the new
methods. While reviewing how to address the underlying cause with a
clippy fix it became obvious a few other methods were missing, too.

197g committed:
Split ImageFile off ImageReader

The latter interface will be something that interfaces and interacts
with the underlying decoder. Introduces ImageDecoder::set_viewport as a
stand-in for what is intended here: color, subsampling, gain map
application will take more interaction, clarification of how the
library will request their data, and adjustments to the layout
definition by the decoder.

197g committed:
Split CMS to private trait interface, icc_profile

The purpose of the trait is to be careful with the boundary to the
`moxcms` dependency. We use it because it is a quality implementation
but it is heavy weight for what we need. There's other possible ways to
provide transfer functions and color space transforms. Now this also
introduces ICC profile parsing but again that could be done with a
*much* lighter dependency as we only need basic information from it. The
trait should make every little additional cross-dependency a conscious
decision.

Also it should be the start of a customization point, by feature flag or
actually at runtime.

197g committed:
Provide limited metadata decoding with ImageReader

fintelia commented on src/io/image_reader_type.rs:
```diff
@@ -354,10 +374,85 @@ impl ImageReader<BufReader<File>> {
+        self.limits.check_dimensions(layout.width, layout.height)?;
+        // Check that we do not allocate a bigger buffer than we are allowed to
+        // FIXME: should this rather go in `DynamicImage::from_decoder` somehow?
+        self.limits.reserve(layout.total_bytes())?;
```
This has a subtle behavior change. Previously we reserved the space for the output buffer and _then_ passed the reduced memory limit onto the decoder to limit its metadata/scratch buffer allocations. But at this point in the code we've already told the decoder that it is free to use the entire memory limit amount for those purposes. So we effectively only limiting memory to 2x the requested amount.

197g committed:
Provide ImageReader::skip

197g committed:
Rename ImageFile -> ImageReaderOptions

197g committed:
Add more_images, ImageAttributes for animation

This allows us to write a generic iterator which uses the same decoder
function to generate a whole sequence of frames. The attributes are
designed to be extensible to describe changes in available metadata as
well, very concretely some formats require that XMP/ICC/… are polled for
each individual image whereas others have one for the whole file and put
that at the end. So there's no universal sequence for querying the
metadata, and we need to hold runtime information. This will be the
focus of the next commit.

197g committed:
Fixup gif testing

No longer responsible for ensuring the size constraints are met under
the new policy and with available of constructing a reader from an
instance of a boxed decoder.

197g committed:
Adjust tests for ImageReader limits usage

197g committed:
Add metadata sequence indicators

fintelia commented:
> @fintelia I understand this is too big for a code-depth review but I'd be interested in the directional input. Is the merging of 'animations' and simple images as well as the optimization hint methods convincing enough?

I do like having animation decoding also handled by `ImageDecoder`. I'm a bit torn about whether it should be a single method or having separate `read_image` and `next_frame` methods, mostly so that PNG's thumbnail can be extracted separately from the frames of an APNG animation.

We should probably also have an `is_animated` method to indicate whether there's multiple animation frames. And perhaps an `is_image_sequence` (open to other names) to support TIFF image sequences. 

I also like the `viewbox` approach, and it is a good demonstration of why this PR is useful. However, it now seems like a good candidate to spin out into a separate PR. The mistake we made with the decode-rect API was not having implementations for very many formats and then discovering that it was too much effort to add them later. So to avoid repeating that, we should ideally include implementations it for a few non-trivial formats when adding it.

> Is the idea of returning data from read_image something that works for you? The struct is meant to be Default-able and fills in the information for into_frames() but I'll sketch out some way of putting the metadata indicators in there (i.e. should you poll xmp for this frame, or wait until the end, or is it constant of the file).

I think it is a bit awkward that you have to decode the image data to know whether the metadata could appear after it. My understanding is that it is mostly for PNG to say that the metadata could come late in the file? (The WebP and TIFF formats also can have metadata after the pixel data, but they provide file offsets so our decoders directly seek to/from that file position)

For PNG we're in an especially weird position where web browsers will generally ignore metadata after the image data, while certain other viewers load it normally. Picking either option means someone is going to think we're doing it wrong. My inclination is that the default should be matching web browsers with an API to opt-in to getting trailing metadata, but I don't know what that interface would look like.

Frame delay could also be a separate method on `ImageDecoder`, but it is probably fine either way.

### Other thoughts

* I think the `ImageReader` type should provide convenience methods for getting the color type, dimensions, and perhaps even width/height individually.
* I'm not totally sold on `ImageDecoder::peek_layout` over separate `dimensions`/`color_type`/etc. methods, but it is probably fine for the lower level API.
* We should remember to have `ImageReader::decode` query/apply the image orientation and as a possible addition maybe have a `decode_raw` method that skips applying any metadata.

197g commented:
> I also like the viewbox approach, and it is a good demonstration of why this PR is useful. 

Great to hear, I'll add a commit that removes it so it can be reverted as a basis of building it more fleshed out. It should be easy to provide more API surface in `tiff` for this and then the performance benefits can be demonstrated.

> I think it is a bit awkward that you have to decode the image data to know whether the metadata could appear after it. My understanding is that it is mostly for PNG to say that the metadata could come late in the file? (The WebP and TIFF formats also can have metadata after the pixel data, but they provide file offsets so our decoders directly seek to/from that file position)

I wouldn't say mostly, the other point is that TIFF has metadata _per image_ and advancing to the next would be destructive in the sense that one would skip one set of possibly available metadata. With hindsight bias of the structure in this PR, the case where it is delayed beyond all image data in IPTC-for-PNG (other kinds of metadata are sane) is less of a problem in my eyes since it at least behaves consistently for both single images and sequences.

- [ ] That said, those fields can clearly be moved. Is it constant for the whole decoder? In that case my best idea is creating a struct for these which also contains the other information, `is_animated` and `is_image_sequence` (though not quite sure what to expect from these semantically, but definitely discourage a check for `more_images` in `into_frames`). Then `ImageReader` should acquire these once which also gets rid of the wart of `last_attributes` being a bad buffer. And should anything pop up where they are not constants then adding a `VariesPerImage` variant seems straightforward (independent of where that poll-each information should live).

> My inclination is that the default should be matching web browsers with an API to opt-in to getting trailing metadata, but I don't know what that interface would look like.

For metadata only available on finish, `ImageReader` could offer such a utility that would either consume the reader after a call to `ImageDecoder::finish` and the metadata read, or add a `seek` method (modelled after tiff) first with semantics of ensuring they can still be used after `finish`.

## Other thoughts

* I think the ImageReader type should provide convenience methods for getting the color type, dimensions, and perhaps even width/height individually. :+1: will do when removing them from the trait.
* I'm not totally sold on ImageDecoder::peek_layout over separate dimensions/color_type/etc. methods, but it is probably fine for the lower level API: Right, that was the idea. Also the decoders I looked at, internally, have a clean sequence point where all that information is available and would poll until there for *either*. So it didn't really make sense to split them up too much. (Also it gives us one clean non_exhaustive struct as an extension point in the sequence should something come up).
* We should remember to have ImageReader::decode query/apply the image orientation and as a possible addition maybe have a decode_raw method that skips applying any metadata.

197g committed:
Move metadata hints into decoder attributes

197g committed:
Remove ImageDecoder::viewbox demonstration

This commit is meant to be revertable. The demonstration code was an
experiment at improving the protocol between the trait implementation
and the user facing `ImageReader`. While successfully showing that the
interface would work with extension points as intended the details are
not fleshed out and we would like at least some implementations to
provide actual `viewbox` internals to add this feature.

Also remember: this is not yet tested, `ImageReader::peek_layout` and
all its derived methods should very likely consult the viewbox setting.

197g committed:
Refactor ImageDecoder::orientation

This method was mostly implemented to acquire EXIF information and
retrieve the entry from it. However, that is a weird way of using the
trait system with this mutable method. Instead we move the
responsibility of extracting and manipulation of EXIF to the ImageReader
and have the decoder interface only support its input. The orientation
is added as a field to per-image decoded data in case where such
information exists outside of EXIF (e.g. in TIFF).

197g commented on src/io/image_reader_type.rs:
```diff
@@ -336,10 +357,85 @@ impl ImageReader<BufReader<File>> {
+        self.viewbox = Some(viewbox);
+    }
+
+    /// Get the previously decoded EXIF metadata if any.
```
Moving the metadata query between `peek_layout` and `read_image` doesn't really affect this argument with the variants of `MetadataHint` that I've found to be necessary. So that is cleaner, see https://github.com/image-rs/image/pull/2672#issuecomment-3703780868

Of course we have `last_attributes` still remaining since that is information that combines the information the reader has retrieved, e.g. the orientation given directly from `read_image` together with the fallback of querying it from exif.

197g committed:
Move metadata between peek_layout, read_image

At least the per-image metadata should be retrieved there. This also
ensures that there is a consistent location for data available after the
header for all images and data that changes per image. The call to
`read_image` can then be destructive for all metadata that is only valid
for that particular frame (e.g. TIFF) which it needed to delay until the
next `peek_layout` before, leading to very odd interactions. Previously,
it meant that `read_image` was supposed to cleanup the image data itself
but not the metadata. This split responsibility is not super intuitive
for many image types..

197g committed:
Replace ImageDecoder::{dimensions,color_type}

Having separate methods for these is not very effective in terms of
protocol. All the formats we have can determine them at the same
sequence point and `peek_layout` has become somewhat central in the
definition of the protocol. It defines the buffer requirements for
`read_image`—having one method means one source of truth for this length
requirement which is far clearer than requiring that the caller run
multiple methods and combine their results (with unclear semantics of
running intermediate other methods).

197g commented:
> I do like having animation decoding also handled by ImageDecoder. I'm a bit torn about whether it should be a single method or having separate read_image and next_frame methods, mostly so that PNG's thumbnail can be extracted separately from the frames of an APNG animation.

I think that's the main remaining interface question. My take on it would be to put this information into `DecodedImageAttributes` together with `delay`. Since the whole struct is a non_exhaustive bag there should be ample room for a 'relation'-kind of information, including alternates for thumbnails, identifiers for frames you can seek back to, etc. How exactly should probably refer to some prior art. (And for that reason, acting similar to a browser makes a lot of sense in other regards, too).

@fintelia  What's not entirely clear to me is if we should turn such information into a separate call, like metadata, between `peek_layout` and `read_image` or if it should get attached to the return value from `read_image`. But given this is somewhat minor compared to the overall structural impact of the other part of this rework, can we move this question to a secondary PR? I'll add comments regarding this to the V1 issue and relevant code.

ronjakoi commented:
I think it would be nice to have an iterator of all images in a file, where in the case of pyramid TIFFs for example, I could filter for the image with dimensions most closely matching a given need.

197g committed:
Remove AnimationDecoder trait

mstoeckl commented on src/io/decoder.rs:
```diff
@@ -1,21 +1,105 @@
+    ///
+    /// The layout returned by an implementation of [`ImageDecoder::peek_layout`] must match the
+    /// buffer expected in [`ImageDecoder::read_image`].
+    fn peek_layout(&mut self) -> ImageResult<crate::ImageLayout>;
```
To better understand how the new ImageDecoder interface works, I've been trying to implement it for the `image-extras` decoders, and as time permits will add comments as I run into issues and have questions.

1. The name ("peek_...") suggests that it is safe to call this multiple times in a row; for example: that the sequence `self.peek_layout()`, `self.peek_layout()`, `self.read_image()` should work and would not read two headers from the input. Could this be explicitly documented?
2. For formats containing a single image, may and/or should calls to peek_layout() return an Error (like ParameterErrorKind::NoMoreData) after read_image() has been called?

RunDevelopment commented on tests/limits.rs:
```diff
@@ -51,7 +51,10 @@ fn permissive_limits() -> Limits {
     limits.max_image_width = Some(WIDTH);
     limits.max_image_height = Some(HEIGHT);
-    limits.max_alloc = Some((WIDTH * HEIGHT * 5).into()); // `* 3`` would be an exact fit for RGB; `* 5`` allows some slack space
+    // `* 3`` would be an exact fit for RGB;
```
```suggestion
    // `* 3` would be an exact fit for RGB;
```

RunDevelopment commented on src/io/decoder.rs:
```diff
@@ -1,21 +1,105 @@
+    /// You should check [`DecoderAttributes::is_animated`] before calling this method. It will
+    /// only be available on animated images. Additionally, most file formats store the metadata in
+    /// the header which might not be read until after calling [`ImageDecoder::peek_layout`].
+    fn animation_attributes(&mut self) -> Option<DecodedAnimationAttributes> {
```
1. What does "available" mean here? Does that mean that `is_animated == false` implies `animation_attrs() == None`?

    If so, then I would suggest replacing the first two sentences with: "This method always returns `None` for non-animated images. Use \[\`DecoderAttributes::is_animated\`\] to check whether an image is animated."

2. Why is this method `mut`? `Self::attributes` isn't `mut` and it's not clear to me why this method would change the internal state of the decoder.

RunDevelopment replied:
> In the sense that the interface is not `unsafe` that is always okay.

I was more so asking about what is correct in terms of the interface. So would we consider behavior like return values changing a bug if a decoder behaved that way?

RunDevelopment reviewed (commented):
I love how this came along since the last time I had a look. Awesome work, @197g and everyone involved!

197g committed:
Move original_color_type into ImageLayout

This makes the pain point mentioned in a review comment more apparent
for us, non_exhaustive is harder to construct but we may need the
extensibility for per-frame control over the decoding process. (Two
likely examples: fields that indicate existing HDR gain maps and
information about the color conversion chain to undo in encoding).

197g committed:
Add ImageReader metadata access adapter

Calling it DecodedImageAttributes so it has the right prefix binding it
to the decoder but clarifies it is not the image data itself. It lives
in the metadata module as you're mostly expected to use it by the
construction from ImageReader and not directly as a type.

197g committed:
Rename DecoderAttributes for clarity

197g commented:
Sorry for all the force-push/rebase confusion. I think I accidentally bisected during a rebase and somehow got two commits squashed in the process of rebasing on the diverged main branch while adding my own work (and still wanting good CI for the pure rebase).

197g committed:
Move original_color_type to DecodedImageAttributes

Motivated by attempting integration with wondermagick. This is part of
the metadata group available after decoding and does, by definition, not
influence the layout. This placement also makes it impossible to be
interpreted that way. In the future the decoder may return a chain of
transformations that it undertook, this being (part of) the base state.
This whole chain would obviously only be available afterwards.

197g reviewed (commented):
Continued as a comment since it's easier to read. @Shnatsel

197g committed:
Return metadata from ImageReader::decode

197g committed:
Merge remote-tracking branch 'origin/main' into decoder-metadata-interface

Shnatsel reviewed (commented):
The shape of the public API looks good now. I left some nits but they're minor.

There's no option to decode without metadata, but given that you need metadata to correctly display an image anyway (Exif for orientation, ICC for color profiles), I think it's fair not to provide such an API.

What happens if decoding metadata returns an error? Does the whole decoding error out, do we silently ignore it and keep going with decoding the pixels, or something else? Do the decoding plugins have to do anything to match the desired behavior or does `image` handle it for them?

197g committed:
Document ImageReader construction / iteration

197g commented on src/io/image_reader_type.rs:
```diff
@@ -369,11 +421,421 @@ impl ImageReader<BufReader<File>> {
+    ///
+    /// Polls the underlying decoder for any `InHeader` metadata that is constant across a file,
+    /// applicable to all images, and appears early.
+    fn fill_header_metadata_if_any(&mut self) -> Result<(), ImageError> {
```
> What happens if decoding metadata returns an error?

Two things: if fetching it goes wrong you get an error. The decoder can also return `UnsupportedError` which will get it silently ignored, this will let us add partial support for forms of metadata to the pipeline of `decode` without fear of breaking images that previously decoded fine.

I suppose there should be as well proxy methods for `ImageDecoder::*_metadata` that are always loud. Access with the metadata adapter will retry in-header metadata if the buffer is still `None`. For in-header metadata there could a method directly on `ImageDecoder` that does the equivalent of `fill_header_metadata_if_any` but without reading any image.

197g committed:
Touch-up for ImageDecoder/ImageReader docs

197g commented:
We could use the strictness configuration for this? If you think `SpecCompliance::Lenient` is the right way to put that (which is also the default).

197g committed:
Delay all metadata errors

197g committed:
Merge remote-tracking branch 'origin/main' into decoder-metadata-interface

fintelia commented on src/io.rs:
```diff
@@ -31,3 +38,66 @@ impl<R: io::Read> ReadExt for R {
+    /// A layout matching a [`DynamicImage`][`crate::DynamicImage`] of the given dimensions and
+    /// color type.
+    pub fn new(w: u32, h: u32, color: crate::ColorType) -> ImageLayout {
+        #[allow(deprecated)]
```
What's deprecated here?

fintelia commented on src/io/decoder.rs:
```diff
@@ -1,24 +1,109 @@
+    /// decoder, e.g. when metadata is available and when it will be overridden. It also provides
+    /// basic capability information about the format. If, in the future, we added different basic
+    /// methods of retrieving color data then the attributes would indicate the preferred and/or
+    /// possible choices.
```
nits:

```suggestion
    /// Retrieve general information about the decoder / the format itself.
    ///
    /// This hints which methods should be called while decoding (a sequence of) images from this
    /// decoder, e.g. when metadata is available and when it will be overwritten. It also provides
    /// basic capability information about the format. If, in the future, we added different basic
    /// methods of retrieving color data then the attributes would indicate the preferred and/or
    /// possible choices.
```

fintelia commented on src/io/image_reader_type.rs:
```diff
@@ -369,11 +476,595 @@ impl ImageReader<BufReader<File>> {
+        // Determine which orientation to use.
+        if attr.orientation.is_none() {
+            attr.orientation = exif.and_then(|chunk| Orientation::from_exif_chunk(&chunk));
+        }
```
Shouldn't this happen within the individual decoders? It seems confusing to have `read_image` have a field to return the orientation but sometimes have it be wrong.

fintelia commented on src/io/image_reader_type.rs:
```diff
@@ -369,11 +476,595 @@ impl ImageReader<BufReader<File>> {
+    /// [`ImageDecoder::read_image`] with kind set to [`None`](crate::io::SequenceControl::None),
+    /// which is also treated as end of stream. This may be used by decoders which can not
+    /// determine the number of images in advance.
+    pub fn into_frames(mut self) -> Frames<'stream> {
```
We should be clear about whether this also applies to image sequences

fintelia commented on src/io/image_reader_type.rs:
```diff
@@ -347,7 +349,112 @@ impl<'a, R: 'a + BufRead + Seek> ImageReader<R> {
+    inner: &'reader mut (dyn ImageDecoder + 'reader),
+    attributes: &'reader DecodedImageAttributes,
+    metadata_buffers: &'reader mut MetadataBuffers,
+}
```
I haven't had a chance to think it through in detail, but it might make sense to have this be a flat struct containing the metadata:

```rust
pub struct DecodeImageMetadata {
    pub orientation: Option<Vec<u8>>,
    pub exif: Option<Vec<u8>>,
    ...
}

fintelia reviewed (commented):
Left a bunch of comments. Haven't had a chance to fully read/consider the `image_reader_type.rs` changes, but I like the direction this is going

197g commented on src/io/decoder.rs:
```diff
@@ -1,24 +1,109 @@
+    ///
+    /// This must be called before a call to [`Self::read_image`] to ensure that the initial
+    /// metadata has been read. In contrast to a constructor it can be called after configuring
+    /// limits and context which avoids resource issues for formats that buffer metadata.
```
I think there should be _some_ method here that necessarily precedes the other metadata calls because it makes the contract rather easy to describe. It may be awkward to surface all kinds of errors in the metadata methods especially considering that the other kinds of metadata (may) get silent failure models. Having this method allows better error ergonomics, I hope.

And the basic layout requirements are always convenient to have, so in terms of Api ergonomics it seems prudent to just include them. All that said, maybe we should not return `ImageLayout` but another wrapper here. If we ever add fields intended for the decoder to communicate per-image analogues of `format_attributes` (e.g. information about _how_ to negotiate decoding color conversion) then it would be odd to add them to `ImageLayout`.

197g replied:
One good alternative name for this might be `prepare_image`. The documentation would then have us use the following sequence of calls:

```
/// ```text,bnf
/// decoding sequence = configure, { decode image }, "finish", { metadata }
///     
/// decode image =
///    "prepare_image", { metadata | "prepare_image" }, "read_image"
///
/// configure = "set_limits"
///
/// metadata = "xmp_metadata" | "icc_profile" | "exif_metadata" | "iptc_metadata"
/// ```
```

Also I'd then move `ImageLayout` into a `layout` field of a `DecoderPreparedImage` struct and if we need to communicate more data than the layout itself we'd extend the latter rather than the former. (I do want `ImageLayout` to describe the shape of `DynamicImage` in any case; we're dearly missing that for a bunch of APIs for instance passing multiple parameters in the encoder).

197g committed:
Resolve naming and structure nits

197g committed:
Turn decode_into to a byte-slice API

197g committed:
Rename prepare_layout, as DecoderPreparedImage

fintelia commented on src/io/image_reader_type.rs:
````diff
@@ -369,11 +469,662 @@ impl ImageReader<BufReader<File>> {
+    /// ```
+    pub fn decode_into(&mut self, buffer: &mut [u8]) -> ImageResult<DecodedImageMetadata<'_>> {
+        let layout = self.inner.peek_layout()?;
+        self.fill_header_metadata_if_any();
````
I've thought about this more, and I don't think it is reasonable to say that info like the image color space or the orientation might just not be available when decoding individual frames. Incremental frame at-a-time decoding isn't very useful if we can get to the end and then say "oh, by the way, make sure to rotate the animation before displaying it". If we do that, users are effectively required to buffer the entire animation in memory before they can display it.

Especially since most users are going to be operating on a byte slice or a `File` object, both of which easily allow jumping back and forth within the file.

197g replied:
Alright that is very reasonable, I've removed `AfterFinish` then. This will require patches to `gif`, `png`, `webp` but that's just a bug in the decoder as currently.

197g committed:
Remove AfterFinish metadata hint

Consolidates the variants so that all supported types of metadata must
guarantee that the data is actually present. This merely requires the
decoder to be able to seek; which is already usually the case and
reasonably implementable.

This will require support in: gif, png, webp

197g committed:
Merge remote-tracking branch 'origin/main' into decoder-metadata-interface

197g committed:
Merge remote-tracking branch 'origin/main' into decoder-metadata-interface

197g committed:
Merge remote-tracking branch 'origin/main' into decoder-metadata-interface

197g committed:
Merge remote-tracking branch 'origin/main' into decoder-metadata-interface

197g committed:
Merge remote-tracking branch 'origin/main' into decoder-metadata-interface

RunDevelopment commented:
Awesome to see this land! 🎉
